### ISSUE TITLE
Frame <= in terms of < or == and >= in terms of > or ==

### ISSUE THREAD

**Issue by glyn** [ietf-wg-jsonpath/draft-ietf-jsonpath-base#245]
According to the current draft certain types, such as Booleans, are unordered:
* equality (`==`) and inequality (`!=`) of values of these types are specified
* ordered comparisons (`<`, `<=`, `>`, and `>=`) are specified to yield false for values of these types.

However, `<=` is often understood the way it is read: "less than or equal to". Similarly for `>=`.

If we could define `<=` as "`<` or `==`" and `>=` as "`>` or `==`", then they could be easier to understand.

With this change `true <= true`, for example, would yield true (whereas according to the current draft, it yields false).

## Total orders

There is often an assumption that orderings are "total" or "linear" (see wikipedia for a [definition](https://en.wikipedia.org/wiki/Total_order)). A total order obeys the following laws:
```
a < b if and only if not b <= a
a > b if and only if not b >= a
```

But according to the current draft, for example,  `false < true` and `true <= false` both yield false, contrary to the first of the above laws. This is because the ordering of Boolean values is _not_ total.

In the discussion at today's interim meeting (2022-08-17), it was suggested that the above laws could cause a problem for this issue. This does not appear to be the case because we are not dealing with total orders.

**Comment by cabo** [comment]
The above section about </>= is phrased in terms of total orders; I believe this is a POLS violation as most people will not think in terms of total orders here.

**Comment by glyn** [comment]
@cabo Ok. That section isn't necessary: I was merely trying to capture the concern you expressed in the meeting. Please could you explain your concern with the above proposal (ignoring the "Total orders" section)?

**Comment by glyn** [comment]
Note to self: to avoid delta clashes, I intend to wait for the following PRs to land or close before submitting a PR for this issue:

- [ ]  https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/pull/241
- [ ] https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/pull/248

**Comment by glyn** [comment]
Note to self (and anyone else who finds mathematics helpful)...

## tl;dr

With the proposed change,  `a <= a` for all `a` _including_ when `a` is "absent", an object, a boolean, or `null`.

Apart from that `<=` is just as in the current draft.

Similarly for `>=` which can be dealt with similarly.

## Reflexive closure

The ordering `<=` in the current draft is a _partial order_, that is, the following are true for all a, b, c:
```
a <= b (reflexive)
if a <= b and b <= c then a <= c (transitive)
if a <= b and b <= a then a == b (antisymmetric)
```

When restricted to numbers or restricted to strings `<=` is a _total order_. That is if `a` and `b` are both numbers or both strings, then in addition to the partial order axioms, the following is true:
```
a <= b or b <= a (strongly connected)
```

Note: `<=` is _not_ a total order when restricted to arrays (if we decide to support `<=` for arrays), since, for example:
```
not [true] <= [false], and
not [false] <= [true]
```

The ordering `<` in the current draft is a _strict partial order_:
```
not a < a  (irreflexive)
if a < b then not b < a (asymmetric)
if a < b and b < c then a < c (transitive)
```

The [_reflexive closure_](https://en.wikipedia.org/wiki/Partially_ordered_set#Correspondence_of_strict_and_non-strict_partial_order_relations) of `<`  _defines_ `<=` _in terms of_ `<`:
```
a <= b if and only if (a < b or a == b)
```

The result is the same as in the current draft except that `a <= a` for all `a`. So, for example, `$.absent <= $.absent` (where `$.absent` is "absent", since `$.absent == $.absent`), `true <= true`, `null <= null` and `$.obj1 <= $.obj2` (where `$.obj1` and `$.obj2` are objects for which `$.obj1 == $.obj2`).

But `<=` is still a partial order and is still a total order when restricted to numbers or when restricted to strings.

**Comment by glyn** [comment]
NB. if https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/pull/248 lands, `[true] <= [true]` is false, but with the above proposal it would be true.
